bundles/automation/org.eclipse.smarthome.automation.thingsupport/src/main/java/org/eclipse/smarthome/automation/thingsupport/AnnotatedThingActionModuleTypeProvider.java

```diff
@@ -0,0 +1,191 @@
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.automation.thingsupport;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import org.eclipse.smarthome.automation.Action;
+import org.eclipse.smarthome.automation.AnnotatedActions;
+import org.eclipse.smarthome.automation.Module;
+import org.eclipse.smarthome.automation.handler.BaseModuleHandlerFactory;
+import org.eclipse.smarthome.automation.handler.ModuleHandler;
+import org.eclipse.smarthome.automation.handler.ModuleHandlerFactory;
+import org.eclipse.smarthome.automation.module.core.handler.AnnotationActionHandler;
+import org.eclipse.smarthome.automation.module.core.provider.AnnotationActionModuleTypeHelper;
+import org.eclipse.smarthome.automation.module.core.provider.ModuleInformation;
+import org.eclipse.smarthome.automation.type.ActionType;
+import org.eclipse.smarthome.automation.type.ModuleType;
+import org.eclipse.smarthome.automation.type.ModuleTypeProvider;
+import org.eclipse.smarthome.core.common.registry.ProviderChangeListener;
+import org.eclipse.smarthome.core.thing.binding.AnnotatedThingActions;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+
+/**
+ * ModuleTypeProvider that collects actions for {@link ThingHandler}s
+ *
+ * @author Stefan Triller - initial contribution
+ *
+ */
+@Component(service = { ModuleTypeProvider.class, ModuleHandlerFactory.class })
+public class AnnotatedThingActionModuleTypeProvider extends BaseModuleHandlerFactory implements ModuleTypeProvider {
+
+    private final Collection<ProviderChangeListener<ModuleType>> changeListeners = ConcurrentHashMap.newKeySet();
+    private final Map<String, Set<ModuleInformation>> moduleInformation = new ConcurrentHashMap<>();
+    private AnnotationActionModuleTypeHelper helper;
+
+    @Activate
+    protected void activate() {
+        this.helper = new AnnotationActionModuleTypeHelper();
+    }
+
+    @Reference(policy = ReferencePolicy.DYNAMIC, cardinality = ReferenceCardinality.MULTIPLE)
+    public void addAnnotatedThingActions(AnnotatedThingActions annotatedThingActions, Map<String, Object> properties) {
+        Collection<ModuleInformation> moduleInformations = helper.parseAnnotations(annotatedThingActions);
```

Comment:
same here with the `helper`.

Reply:
I'm not 100% sure, but: I would tend to call all this code after `activate` was called. And then again for every `AnnotatedThingActions` instance which gets added here. Same for `deactivate` and also for the "other" service. You might have to add an `activated` flag to distinguish between both states.

Reply:
I am also not 100% sure about this point. The `ProviderChangelistener` are not injected via OSGi, so i am not sure whether I have to wait until the service is activated before calles this code.

@kaikreuzer I would appreciate your feedback on this.

Reply:
Afaics, the logic here does NOT in any way depend on / require stuff that is initialized in `activate` nor any injected service. If this is correct, I agree that the logic can stay here. if not, it should be done the way @htreu suggests.

Reply:
Yes, agreed. With `helper` always available this should stay as it is.